src/twisted/internet/abstract.py

```diff
@@ -19,7 +19,7 @@
 
 
 def _dataMustBeBytes(obj):
```

Comment:
so FileDescriptor.write calls `self._tempDataBuffer.append(data)` which is used in `doWrite` which calls:

```python
self.dataBuffer = _concatenate(
    self.dataBuffer, self.offset, self._tempDataBuffer
)
```

Which calls `bytes.join` which is defined as taking a `typing.ByteString | memoryview`:
https://github.com/python/typeshed/blob/16ae4c61201cd8b96b8b22cdfb2ab9e89ba5bcf2/stdlib/builtins.pyi#L440

So maybe we should be type-checking (at runtime and with mypy) for that instead of bytes

Reply:
looks like this is still wrong: https://github.com/python/typing/issues/593#issuecomment-605516078

Reply:
I think we want: https://github.com/python/typeshed/blob/c92e3cb228d478a4cbd8f53072cae608dcf2f6c4/stdlib/_typeshed/__init__.pyi#L171

```python
ReadableBuffer = Union[bytes, bytearray, memoryview, array.array, mmap.mmap]
READABLE_BUFFER = (bytes, bytearray, memoryview, array.array, mmap.mmap)

def _dataMustBeBytes(obj: ReadableBuffer):
    if not isinstance(obj, READABLE_BUFFER):
        raise TypeError("Data must be bytes")
```

and then we need to go through and replace `data: bytes` with `data: ReadableBuffer` everywhere

Reply:
I agree that the protocol should accept any byte-like object (or something that fulfills the buffer protocol); my commits were limited to just supporting `memoryview` but this makes more sense. Just have to expand on the unit tests as well, but I am not sure if twisted's integration with unit tests support sub tests or parameterization.

Reply:
I think that changing the type hint to `ReadableBuffer` as above is good and that removing the `_dataMustBeBytes` entirely would be the best path here…

The type hint will steer people the right way and the type check was there to catch string issues common in Py2, which is no longer in play.

Reply:
Ok so a new public: `twisted.internet.abstract.ReadableBuffer`

```python
ReadableBuffer = Union[bytes, bytearray, memoryview, array.array, mmap.mmap]
```

and _dataMustBeBytes removed with all consumers updated to use the new type annotation.

@wongpat would you like to make these changes?

Reply:
> _dataMustBeBytes removed with all consumers updated to use the new type annotation.

Do we not want to throw an error if someone passes in a unicode value (without calling `encode`)? I'm fine doing as suggested, but wanted to confirm this behavior change.